// @prisma/migrate/dist/commands/DbDrop.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DbDrop_exports = {};
__export(DbDrop_exports, {
  DbDrop: () => DbDrop
});
module.exports = __toCommonJS(DbDrop_exports);
var import_internals = require("@prisma/internals");
var import_chalk = __toESM(require("chalk"));
var import_prompts = __toESM(require("prompts"));
var import_ensureDatabaseExists = require("../utils/ensureDatabaseExists");
var import_errors = require("../utils/errors");
var import_flagErrors = require("../utils/flagErrors");
var import_getSchemaPathAndPrint = require("../utils/getSchemaPathAndPrint");
var import_printDatasource = require("../utils/printDatasource");
const _DbDrop = class {
  static new() {
    return new _DbDrop();
  }
  async parse(argv) {
    const args = (0, import_internals.arg)(argv, {
      "--help": Boolean,
      "-h": "--help",
      "--preview-feature": Boolean,
      "--force": Boolean,
      "-f": "--force",
      "--schema": String,
      "--telemetry-information": String
    });
    if ((0, import_internals.isError)(args)) {
      return this.help(args.message);
    }
    await (0, import_internals.checkUnsupportedDataProxy)("db drop", args, true);
    if (args["--help"]) {
      return this.help();
    }
    if (!args["--preview-feature"]) {
      throw new import_flagErrors.PreviewFlagError();
    }
    (0, import_internals.loadEnvFile)(args["--schema"], true);
    const schemaPath = await (0, import_getSchemaPathAndPrint.getSchemaPathAndPrint)(args["--schema"]);
    const datasourceInfo = await (0, import_ensureDatabaseExists.getDatasourceInfo)({ schemaPath, throwIfEnvError: true });
    (0, import_printDatasource.printDatasource)({ datasourceInfo });
    const schemaDir = await (0, import_internals.getSchemaDir)(schemaPath);
    console.info();
    if (!args["--force"]) {
      if (!(0, import_internals.canPrompt)()) {
        throw new import_errors.DbNeedsForceError("drop");
      }
      const confirmation = await (0, import_prompts.default)({
        type: "text",
        name: "value",
        message: `Enter the ${datasourceInfo.prettyProvider} database name "${datasourceInfo.dbName}" to drop it.
Location: "${datasourceInfo.dbLocation}".
${import_chalk.default.red("All data will be lost")}.`
      });
      console.info();
      if (!confirmation.value) {
        console.info("Drop cancelled.");
        process.exit(130);
      } else if (confirmation.value !== datasourceInfo.dbName) {
        throw Error(`The database name entered "${confirmation.value}" doesn't match "${datasourceInfo.dbName}".`);
      }
    }
    if (await (0, import_internals.dropDatabase)(datasourceInfo.url, schemaDir)) {
      return `${process.platform === "win32" ? "" : "\u{1F680}  "}The ${datasourceInfo.prettyProvider} database "${datasourceInfo.dbName}" from "${datasourceInfo.dbLocation}" was successfully dropped.
`;
    } else {
      return "";
    }
  }
  help(error) {
    if (error) {
      return new import_internals.HelpError(`
${import_chalk.default.bold.red(`!`)} ${error}
${_DbDrop.help}`);
    }
    return _DbDrop.help;
  }
};
let DbDrop = _DbDrop;
__name(DbDrop, "DbDrop");
DbDrop.help = (0, import_internals.format)(`
${process.platform === "win32" ? "" : import_chalk.default.bold("\u{1F4A3}  ")}Drop the database

${import_chalk.default.bold.yellow("WARNING")} ${import_chalk.default.bold(
  `Prisma db drop is currently in Preview (${(0, import_internals.link)("https://pris.ly/d/preview")}).
There may be bugs and it's not recommended to use it in production environments.`
)}
${import_chalk.default.dim("When using any of the subcommands below you need to explicitly opt-in via the --preview-feature flag.")}

${import_chalk.default.bold("Usage")}

  ${import_chalk.default.dim("$")} prisma db drop [options] --preview-feature

${import_chalk.default.bold("Options")}

   -h, --help   Display this help message
     --schema   Custom path to your Prisma schema
  -f, --force   Skip the confirmation prompt

${import_chalk.default.bold("Examples")}

  Drop the database
  ${import_chalk.default.dim("$")} prisma db drop --preview-feature

  Specify a schema
  ${import_chalk.default.dim("$")} prisma db drop --preview-feature --schema=./schema.prisma

  Use --force to skip the confirmation prompt
  ${import_chalk.default.dim("$")} prisma db drop --preview-feature --force
`);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DbDrop
});
